### PR TITLE
Add a minimal tracer example

### PR DESCRIPTION
**Pull request by pyohannes**
This adds a trivial example, which sets up a tracer in the main application and uses a library that is instrumented with OT. The example tries to be as minimal as possible (the biggest chunk being a stdout exporter).

I had to move some other bits to make the example work:
* The `simple_processor.h` needs to be moved to the public include directory to be available for application developers to use.
* The tracer in the tracer provider needs to be stored as in a `shared_ptr` instead of a `nostd::shared_ptr`, otherwise we run into problems with `shared_from_this`.

Some things still need to be done: timestamps, durations and ids are not initialized for spans.  I'll probably look into this in a separate PR.

### CODE REVIEW

File: examples/simple/BUILD [hunk 1]
```diff
@@ -0,0 +1,12 @@
+cc_binary(
+    name = "example_simple",
+    srcs = [
+        "library.cc",
```
Comment (g-easy) [inline examples/simple/BUILD:4]: Please consider putting this into a separate cc_library target.

File: examples/simple/library.cc [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#include "opentelemetry/trace/provider.h"
+
+namespace trace = opentelemetry::trace;
+namespace nostd = opentelemetry::nostd;
+
+static nostd::shared_ptr<trace::Tracer> get_tracer()
```
Comment (g-easy) [inline examples/simple/library.cc:6]: Optional: wrap in an anonymous namespace instead of "static"

File: examples/simple/library.cc [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+{
+  auto provider = trace::Provider::GetTracerProvider();
+  return provider->GetTracer("");
```
Comment (g-easy) [inline examples/simple/library.cc:9]: People will copy-paste from examples. Please "" -> "foo_library" or whatever the arg should be. :)

File: examples/simple/library.cc [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+static void f1()
+{
+  auto span = get_tracer()->StartSpan("f1");
+}
```
Comment (g-easy) [inline examples/simple/library.cc:15]: Please add some events. :)
Maybe in a followup PR?

File: examples/simple/main.cc [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#include "opentelemetry/sdk/trace/simple_processor.h"
+#include "opentelemetry/sdk/trace/tracer_provider.h"
+#include "opentelemetry/trace/provider.h"
+
+// Using an exporter that simply dumps span data to stdout.
+#include "stdout_exporter.h"
+
+// This models a call to an external library that is instrumented with OT.
+extern void library();
```
Comment (g-easy) [inline examples/simple/main.cc:9]: 1. Doesn't need extern.

2. Please move this line to library.h, so it feels more like a cc_library for bazel people.

File: examples/simple/main.cc [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+void initTracer()
```
Comment (g-easy) [inline examples/simple/main.cc:11]: Anonymous namespace.

File: examples/simple/main.cc [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+{
+  auto exporter  = std::unique_ptr<sdktrace::SpanExporter>(new StdoutExporter);
+  auto processor = std::shared_ptr<sdktrace::SpanProcessor>(
+      new sdktrace::SimpleSpanProcessor(std::move(exporter)));
+  auto provider = nostd::shared_ptr<trace::TracerProvider>(new sdktrace::TracerProvider(processor));
```
Comment (g-easy) [inline examples/simple/main.cc:16]: std::move(processor) ?

File: examples/simple/main.cc [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  trace::Provider::SetTracerProvider(provider);
```
Comment (g-easy) [inline examples/simple/main.cc:17]: Can/should this also be std::moved?

File: examples/simple/stdout_exporter.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+#include "opentelemetry/sdk/trace/exporter.h"
+#include "opentelemetry/sdk/trace/span_data.h"
+
+#include <iostream>
+
+namespace trace    = opentelemetry::trace;
+namespace nostd    = opentelemetry::nostd;
+namespace sdktrace = opentelemetry::sdk::trace;
+
+class StdoutExporter final : public sdktrace::SpanExporter
```
Comment (g-easy) [inline examples/simple/stdout_exporter.h:12]: Probably not to be resolved in this PR but: should we have a canonical StdoutExporter in the SDK?

File: examples/simple/stdout_exporter.h [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+{
+  std::unique_ptr<sdktrace::Recordable> MakeRecordable() noexcept
+  {
+    return std::unique_ptr<sdktrace::Recordable>(new sdktrace::SpanData);
+  }
+
+  sdktrace::ExportResult Export(nostd::span<std::unique_ptr<sdktrace::Recordable>> &spans) noexcept
```
Comment (g-easy) [inline examples/simple/stdout_exporter.h:19]: How come none of this is const?

File: examples/simple/stdout_exporter.h [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        char trace_id[32]       = {0};
+        char span_id[16]        = {0};
+        char parent_span_id[16] = {0};
+
+        span->GetTraceId().ToLowerBase16(trace_id);
+        span->GetSpanId().ToLowerBase16(span_id);
+        span->GetParentSpanId().ToLowerBase16(parent_span_id);
+
+        std::cout << "{" << std::endl
+                  << "  name          : " << span->GetName() << std::endl
```
Comment (g-easy) [inline examples/simple/stdout_exporter.h:37]: We previously decided "\n" over endl.

File: examples/simple/stdout_exporter.h (hunk 9 above)
Comment (pyohannes) [reply examples/simple/stdout_exporter.h:12]: Yes, I think so. But that should go in a separate PR (than the `StdoutExporter` here can be replaced).

I'll write an issue for that. That could be a good issue for an intern to work on.

File: examples/simple/library.cc (hunk 4 above)
Comment (pyohannes) [reply examples/simple/library.cc:15]: I'll add this in a follow-up PR. Events aren't wired up in the SDK yet.

File: examples/simple/BUILD (hunk 1 above)
Comment (pyohannes) [reply examples/simple/BUILD]: Done.

File: examples/simple/library.cc (hunk 2 above)
Comment (pyohannes) [reply examples/simple/library.cc]: Done.

File: examples/simple/library.cc (hunk 3 above)
Comment (pyohannes) [reply examples/simple/library.cc]: Good call. Done.

File: examples/simple/main.cc (hunk 5 above)
Comment (pyohannes) [reply examples/simple/main.cc]: Done.

File: examples/simple/stdout_exporter.h (hunk 11 above)
Comment (pyohannes) [reply examples/simple/stdout_exporter.h]: Replaced `endl` with `\n`.

File: examples/simple/stdout_exporter.h (hunk 10 above)
Comment (pyohannes) [reply examples/simple/stdout_exporter.h]: I changed the span to `const` (although I think it wouldn't matter much, as the processor sees the content of the span as invalidated anyway after the call to `Export`).

Was this the intent of your question? Or do you also think the recordable itself should be `const`?

**Pull request merged**